<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"/><title>TitanCNA/</title><meta name="viewport" content="width=device-width, initial-scale=1.0"><link rel="alternate" type="application/rss+xml" title="Ha Lab" href="http://gavinha.github.io/feed.xml"><link href="/css/style.css" rel="stylesheet" media="screen"> <script>!function(e){var t,a={kitId:"xeu8jut",scriptTimeout:3e3},c=e.documentElement,i=setTimeout(function(){c.className=c.className.replace(/\bwf-loading\b/g,"")+" wf-inactive"},a.scriptTimeout),n=e.createElement("script"),o=!1,s=e.getElementsByTagName("script")[0];c.className+=" wf-loading",n.src="//use.typekit.net/"+a.kitId+".js",n.async=!0,n.onload=n.onreadystatechange=function(){if(t=this.readyState,!(o||t&&"complete"!=t&&"loaded"!=t)){o=!0,clearTimeout(i);try{Typekit.load(a)}catch(e){}}},s.parentNode.insertBefore(n,s)}(document)</script> <script>!function(e,a,n,t,c,o,s){e.GoogleAnalyticsObject=c,e[c]=e[c]||function(){(e[c].q=e[c].q||[]).push(arguments)},e[c].l=1*new Date,o=a.createElement(n),s=a.getElementsByTagName(n)[0],o.async=1,o.src=t,s.parentNode.insertBefore(o,s)}(window,document,"script","//www.google-analytics.com/analytics.js","ga"),ga("create","UA-26244371-2","bedford.io"),ga("send","pageview")</script> <script src="/js/jquery.min.js"></script> <!link href="css/bootstrap.min.css" rel="stylesheet"> <script src="/js/transition.js"></script><script src="/js/collapse.js"></script><script src="/js/less.min.js"></script></head><body><div id="header"> <nav class="navbar navbar-default navbar-fixed-top" role="navigation"><div class="container"><div class="navbar-header"> <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar-collapse-element" aria-expanded="false"> <span class="sr-only">Toggle navigation</span> <span class="icon-bar"></span> <span class="icon-bar"></span> <span class="icon-bar"></span> </button> <!object class="logo navbar-brand" data="/images/logo.svg" type="image/svg+xml"><!/object> <a class="navbar-brand" href="/">Ha Lab</a></div><div class="collapse navbar-collapse" id="navbar-collapse-element"><ul class="nav navbar-nav navbar-right"><li> <a href="/news/">News</a></li><li> <a href="/research/">Research</a></li><li> <a href="/people/">People</a></li><li> <a href="/papers/">Papers</a></li><li> <a href="/projects/">Software</a></li><li><li><a class="off" href="/join/join/">Join</a></li></li></ul></div></div> </nav></div><div class="container"><div class="hugespacer"></div><div class="bigspacer"></div><div class="row"><div class="col-md-12"><div class="title titlebox"><ol class="breadcrumb"><li>TitanCNA</li></ol></div><p><div class="head"> Analysis of subclonal copy number alterations (CNA) and loss of heterozygosity (LOH) in cancer</div></div></div><div class="bigspacer"></div><div class="row"><div class="col-md-3"><div class="bigspacer"></div><div class="smallhead"> Source code</div><div class="pad-left note"><div class="smallspacer"></div> <i class="fa fa-cog fa-fw"></i> <a class="off" href="https://github.com/gavinha/TitanCNA">github.com/gavinha/TitanCNA</a></div><div class="bigspacer"></div><div class="smallhead"> Contributors</div><div class="pad-left note"><div class="smallspacer"></div><div> <a class="off" href="https://github.com/gavinha"> <img class="pull-left avatar" src="https://avatars1.githubusercontent.com/u/9221688?v=4&s=50"><div class="handlebox" style="padding-left:5px""> gavinha</div> </a></div><div class="smallspacer"></div><div> <a class="off" href="https://github.com/lbeltrame"> <img class="pull-left avatar" src="https://avatars0.githubusercontent.com/u/25192?v=4&s=50"><div class="handlebox" style="padding-left:5px""> lbeltrame</div> </a></div><div class="smallspacer"></div><div> <a class="off" href="https://github.com/dtenenba"> <img class="pull-left avatar" src="https://avatars0.githubusercontent.com/u/2286826?v=4&s=50"><div class="handlebox" style="padding-left:5px""> dtenenba</div> </a></div><div class="smallspacer"></div><div> <a class="off" href="https://github.com/hpages"> <img class="pull-left avatar" src="https://avatars1.githubusercontent.com/u/8810451?v=4&s=50"><div class="handlebox" style="padding-left:5px""> hpages</div> </a></div><div class="smallspacer"></div><div> <a class="off" href="https://github.com/mrjc42"> <img class="pull-left avatar" src="https://avatars3.githubusercontent.com/u/2029097?v=4&s=50"><div class="handlebox" style="padding-left:5px""> mrjc42</div> </a></div></div><div class="bigspacer"></div><div class="smallhead"> Latest commits</div><div class="pad-left smallnote"><ul class="list-unstyled"><div class="smallspacer"></div><li> <i class="fa fa-check-square-o fa-fw"></i> <a class="off" href="https://github.com/gavinha/TitanCNA/commit/e46683c75d28961668d10882294ea0ab53975685"> 26 Jun 2018 - <span class="text-gray">fixed bug in getSubcloneProfiles when sample has no CNA or too noisy #22</span> </a></li><div class="smallspacer"></div><li> <i class="fa fa-check-square-o fa-fw"></i> <a class="off" href="https://github.com/gavinha/TitanCNA/commit/d2687ea377ba2c59005b66fca5fc4fd6a9dec4c6"> 24 May 2018 - <span class="text-gray">fixed bug in selectSolution.R [ci skip]</span> </a></li><div class="smallspacer"></div><li> <i class="fa fa-check-square-o fa-fw"></i> <a class="off" href="https://github.com/gavinha/TitanCNA/commit/1b0d954293435c0fdd51fce769eba01b3164928c"> 11 Apr 2018 - <span class="text-gray">version bump v1.17.2</span> </a></li><div class="smallspacer"></div><li> <i class="fa fa-check-square-o fa-fw"></i> <a class="off" href="https://github.com/gavinha/TitanCNA/commit/7fc4fcc3192aa975a480ed360db0db36d46a6ae6"> 11 Apr 2018 - <span class="text-gray">include function getHaplotypesFromVCF [ci skip]</span> </a></li><div class="smallspacer"></div><li> <i class="fa fa-check-square-o fa-fw"></i> <a class="off" href="https://github.com/gavinha/TitanCNA/commit/635e89acd2fa1b75adca3c7438aa71c8ef0e8586"> 05 Apr 2018 - <span class="text-gray">remove redundant ichorCNA options [ci skip]</span> </a></li></ul></div><div class="spacer"></div><div class="smallhead"> Pages</div><div class="pad-left smallnote"><ul class="list-unstyled"><div class="smallspacer"></div><li> <a class="off" href="/projects/TitanCNA/scripts/R_scripts/"> <i class="fa fa-folder-open-o"></i> / scripts / R_scripts / </a></li><div class="smallspacer"></div><li> <a class="off" href="/projects/TitanCNA/scripts/snakemake/"> <i class="fa fa-folder-open-o"></i> / scripts / snakemake / </a></li><div class="smallspacer"></div><li> <a class="off" href="/projects/TitanCNA/scripts/TenX_scripts/"> <i class="fa fa-folder-open-o"></i> / scripts / TenX_scripts / </a></li><div class="smallspacer"></div><li> <a class="off" href="/projects/TitanCNA/scripts/"> <i class="fa fa-folder-open-o"></i> / scripts / </a></li><div class="smallspacer"></div><li> <a class="off" href="/projects/TitanCNA/"> <i class="fa fa-folder-open-o"></i> / </a></li></ul></div><div class="spacer"></div></div><div class="col-md-8"><div class="post"><p><p><a href="https://travis-ci.org/gavinha/TitanCNA"><img src="https://travis-ci.org/gavinha/TitanCNA.svg?branch=master" alt="Build Status"/></a></p><h1 id="titancna"><em>TitanCNA</em></h1><p>TitanCNA a R/Bioconductor package for analyzing subclonal copy number alterations (CNA) and loss of heterozygosity (LOH) in whole genome and exome sequencing of tumours.</p><p>Ha, G., et al. (2014). <a href="http://genome.cshlp.org/content/24/11/1881">TITAN: Inference of copy number architectures in clonal cell populations from tumour whole genome sequence data. Genome Research, 24: 1881-1893.</a> (PMID: 25060187)</p><h2 id="contact">Contact</h2><p>Gavin Ha<br/> Dana-Farber Cancer Institute<br/> Broad Institute<br/> contact: <a href="mailto:gavinha@gmail.com">gavinha@gmail.com</a> or <a href="mailto:gavinha@broadinstitute.org">gavinha@broadinstitute.org</a><br/> Date: March 22, 2018</p><h2 id="table-of-contents">Table of Contents</h2><ul><li><a href="#links">Links</a></li><li><a href="#news">News</a></li><li><a href="#installation">Installation</a></li><li><a href="#usage">Usage</a></li><li><a href="#vignette-in-titancna-r-package">Vignette in TitanCNA R package</a></li><li><a href="#acknowledgements">Acknowledgements</a></li><li><a href="#software-license">License</a></li></ul><h2 id="links">Links</h2><p>TitanCNA GitHub: https://github.com/gavinha/TitanCNA<br/> Snakemake Workflow: https://github.com/gavinha/TitanCNA/tree/master/scripts/snakemake<br/> Google Groups: https://groups.google.com/forum/#!forum/titancna<br/> TitanCNA website: http://compbio.bccrc.ca/software/titan/<br/> KRONOS TITAN Workflow: https://github.com/MO-BCCRC/titan_workflow<br/> Publication in Genome Research: http://genome.cshlp.org/content/24/11/1881</p><h2 id="news">News</h2><p>(See <a href="NEWS">NEWS</a> for previous version notes)</p><h3 id="titancna-version-1171-changes">TitanCNA version 1.17.1 changes</h3><p>1) New functions: - <code class="highlighter-rouge">correctIntegerCN()</code>: recomputes high-level copy number that is capped by the maximum CN state. Performs two tasks - (1) correct log ratio based on purity and ploidy, and then convert to decimal CN value; (2) Correct bins and segments in which the original predicted integer copy number was assigned the maximum CN state; bins and segments for all of chromosome X are also corrected, if provided in the input.</p><p>2) Modified functions: - <code class="highlighter-rouge">plotSegmentMedians()</code> and <code class="highlighter-rouge">plotCNlogRByChr()</code>: includes argument to show color-coding for corrected copy number; defaults to TRUE for this argument.</p><p>3) Removed functions/manual/dependencies: - <code class="highlighter-rouge">extractAlleleReadCounts()</code> - <code class="highlighter-rouge">Rsamtools</code> dependency</p><h3 id="titancna-version-1150-changes">TitanCNA version 1.15.0 changes</h3><p>1) 10X Genomics analysis</p><ul><li>Please see <a href="scripts/">scripts</a> for instructions on running the 10X Genomics analysis.</li></ul><p>2) New script to help <strong>select optimal solutions</strong>.<br/> Please see <a href="https://github.com/gavinha/TitanCNA/tree/master/scripts/R_scripts">scripts/R_scripts</a></p><p>3) Added snakemake pipeline for entire TITAN workflow<br/> Please see <a href="scripts/snakemake">scripts/snakemake</a>.</p><p>4) New function: - <code class="highlighter-rouge">plotSegmentMedians()</code> - <code class="highlighter-rouge">loadHaplotypeAlleleCounts()</code>: loads input allele counts with phasing information - <code class="highlighter-rouge">plotHaplotypeFraction()</code>: results from 10X Genomics WGS data with phasing of haplotype blocks</p><p>5) Modified features (no changes for user-accessible functions): - updateParameters: coordinate descent estimate of ploidy update uses previously estimated normal parameter from the same corodinate descent iteration ; leads to faster convergence</p><h2 id="installation">Installation</h2><h3 id="install-titancna-r-package-from-github">Install TitanCNA R package from github</h3><p>From within R-3.3.2 or higher,</p><div class="highlighter-rouge"><pre class="highlight"><code>install.packages("devtools")
library(devtools)
install_github("gavinha/TitanCNA")
</code></pre></div><h3 id="install-titancna-from-bioconductor">Install TitanCNA from Bioconductor</h3><p>From within R-3.3.2 or higher,</p><div class="highlighter-rouge"><pre class="highlight"><code>source("https://bioconductor.org/biocLite.R")
biocLite("TitanCNA")
</code></pre></div><h3 id="install-other-dependencies">Install other dependencies</h3><ol><li><p>Install the HMMcopy suite Please follow instructions on the HMMcopy website <a href="http://compbio.bccrc.ca/software/hmmcopy/">http://compbio.bccrc.ca/software/hmmcopy/</a>.</p></li><li><p><a href="https://github.com/MO-BCCRC/titan_workflow">KRONOS</a> TITAN Workflow The easiest way to generate these files is by using the downloadable pipeline from https://github.com/MO-BCCRC/titan_workflow.</p></li></ol><h2 id="usage">Usage</h2><p>R scripts are provided to run the R component of the TITAN analysis using the TitanCNA R/Bioconductor package.<br/> Please go to the <a href="scripts/">scripts</a> directory and look at the README there for more details.</p><h3 id="snakemake-workflow">Snakemake workflow</h3><p>A <a href="scripts/snakemake">snakemake</a> is also provided in this repo.<br/> This workflow will run the TITAN a set of tumour-normal pairs, starting from the BAM files and generating TitanCNA outputs. It will also perform model selection at the end of the workflow to choose the optimal ploidy and clonal cluster solutions.</p><h2 id="vignette-in-titancna-r-package">Vignette in TitanCNA R package</h2><p>The PDF of the vignette can be accessed from R</p><div class="highlighter-rouge"><pre class="highlight"><code>library(TitanCNA)
browseVignettes(package = "TitanCNA")
</code></pre></div><p>The path of the file can also be located using</p><div class="highlighter-rouge"><pre class="highlight"><code>pathToInstall &lt;- system.file(package = "TitanCNA")
pathToPdf &lt;- paste0(pathToInstall, "/int/doc/TitanCNA.pdf)
</code></pre></div><p>The example provided will reproduce Figure 1 in the manuscript. However, it will be slightly different because the example is only based on the analysis of chr2, not genome-wide.</p><h2 id="acknowledgements">Acknowledgements</h2><p>TitanCNA was developed by Gavin Ha while in the laboratories of Sohrab Shah (sshah@bccrc.ca) and Sam Aparicio (saparicio@bccrc.ca) at the Dept of Molecular Oncology, BC Cancer Agency, Vancouver, Canada.<br/> Yikan Wang and Daniel Lai have contributed code and discussions to this project.<br/> The KRONOS TITAN workflow was developed by Diljot Grewal (<a href="mailto:dgrewal@bccrc.ca">dgrewal@bccrc.ca</a>) and Jafar Taghiyar (<a href="mailto:jtaghiyar@bccrc.ca">jtaghiyar@bccrc.ca</a>).<br/> HMMcopy was co-developed by Daniel Lai and Gavin Ha.</p><p>TitanCNA was inspired by existing methods including <a href="https://sites.google.com/site/oncosnp/">OncoSNP</a> and <a href="https://bitbucket.org/aroth85/pyclone/wiki/Home">PyClone</a></p><h2 id="software-license">Software License</h2><p>License: <a href="LICENSE">GPLv3</a></p><p>TitanCNA R code is open source and R/Bioconductor package is under <a href="LICENSE">GPLv3</a>. This applies to the v1.9.0 and all subsequent versions within and obtained from Bioconductor.</p><p>Users who are using TitanCNA earlier than v1.9.0 not for the purpose of academic research should contact gavinha@broadinstitute.org, sshah@bccrc.ca, and prebstein@bccancer.bc.ca to inquire about previous licensing.</p></div></div><div class="col-md-1"></div></div><div class="bigspacer"></div><div class="hugespacer"></div></div><div id="footer"> <nav class="navbar navbar-default navbar-fixed-bottom" role="navigation"><div class="container"><div class="navbar-footer"> <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar-collapse-element" aria-expanded="false"> <span class="sr-only">Toggle navigation</span> <span class="icon-bar"></span> <span class="icon-bar"></span> <span class="icon-bar"></span> </button></div> <!object class="logofooter navbar-brand" data="/images/fhcrc-logo.png" type="image/png"><!/object><div class="collapse navbar-collapse" id="navbar-collapse-element"><ul class="nav navbar-nav navbar-right navbar-tiny"><li><a class="off" href="/misc/about/">About</a></li><li><a class="off" href="/misc/contact/">Contact</a></li></ul><ul class="nav navbar-text"></ul></div></div> </nav></div></body></html>