---
layout: paper
title: "Modeling clonal structure over narrow time frames via circulating tumor DNA in metastatic breast cancer"
authors: Weber ZT, Collier KA, Tallman D, Forman J, Shukla S, Asad S, Rhoades J, Freeman S, Parsons HA, Williams NO, Barroso-Sousa R, Stover EH, Mahdi H, Cibulskis C, Lennon NJ, Ha G, Adalsteinsson VA, Tolaney SM, Stover DG. 
year: 2021
authorshort: Web et al.
journal: Genome Medicine 
vol: 13
page: 89
doi: 10.1186/s13073-021-00895-x
---

# Abstract

Background
Circulating tumor DNA (ctDNA) offers minimally invasive means to repeatedly interrogate tumor genomes, providing opportunities to monitor clonal dynamics induced by metastasis and therapeutic selective pressures. In metastatic cancers, ctDNA profiling allows for simultaneous analysis of both local and distant sites of recurrence. Despite the promise of ctDNA sampling, its utility in real-time genetic monitoring remains largely unexplored.

Methods
In this exploratory analysis, we characterize high-frequency ctDNA sample series collected over narrow time frames from seven patients with metastatic triple-negative breast cancer, each undergoing treatment with Cabozantinib, a multi-tyrosine kinase inhibitor (NCT01738438, https://clinicaltrials.gov/ct2/show/NCT01738438). Applying orthogonal whole exome sequencing, ultra-low pass whole genome sequencing, and 396-gene targeted panel sequencing, we analyzed 42 plasma-derived ctDNA libraries, representing 4–8 samples per patient with 6–42 days between samples. Integrating tumor fraction, copy number, and somatic variant information, we model tumor clonal dynamics, predict neoantigens, and evaluate consistency of genomic information from orthogonal assays.

Results
We measured considerable variation in ctDNA tumor faction in each patient, often conflicting with RECIST imaging response metrics. In orthogonal sequencing, we found high concordance between targeted panel and whole exome sequencing in both variant detection and variant allele frequency estimation (specificity = 95.5%, VAF correlation, r = 0.949), Copy number remained generally stable, despite resolution limitations posed by low tumor fraction. Through modeling, we inferred and tracked distinct clonal populations specific to each patient and built phylogenetic trees revealing alterations in hallmark breast cancer drivers, including TP53, PIK3CA, CDK4, and PTEN. Our modeling revealed varied responses to therapy, with some individuals displaying stable clonal profiles, while others showed signs of substantial expansion or reduction in prevalence, with characteristic alterations of varied literature annotation in relation to the study drug. Finally, we predicted and tracked neoantigen-producing alterations across time, exposing translationally relevant detection patterns.

Conclusions
Despite technical challenges arising from low tumor content, metastatic ctDNA monitoring can aid our understanding of response and progression, while minimizing patient risk and discomfort. In this study, we demonstrate the potential for high-frequency monitoring of evolving genomic features, providing an important step toward scalable, translational genomics for clinical decision making.

