<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"/><title>ichorCNA/scripts/</title><meta name="viewport" content="width=device-width, initial-scale=1.0"><link rel="alternate" type="application/rss+xml" title="Ha Lab" href="http://gavinha.github.io/feed.xml"><link href="/css/style.css" rel="stylesheet" media="screen"> <script>!function(e){var t,a={kitId:"xeu8jut",scriptTimeout:3e3},c=e.documentElement,i=setTimeout(function(){c.className=c.className.replace(/\bwf-loading\b/g,"")+" wf-inactive"},a.scriptTimeout),n=e.createElement("script"),o=!1,s=e.getElementsByTagName("script")[0];c.className+=" wf-loading",n.src="//use.typekit.net/"+a.kitId+".js",n.async=!0,n.onload=n.onreadystatechange=function(){if(t=this.readyState,!(o||t&&"complete"!=t&&"loaded"!=t)){o=!0,clearTimeout(i);try{Typekit.load(a)}catch(e){}}},s.parentNode.insertBefore(n,s)}(document)</script> <script>!function(e,a,n,t,c,o,s){e.GoogleAnalyticsObject=c,e[c]=e[c]||function(){(e[c].q=e[c].q||[]).push(arguments)},e[c].l=1*new Date,o=a.createElement(n),s=a.getElementsByTagName(n)[0],o.async=1,o.src=t,s.parentNode.insertBefore(o,s)}(window,document,"script","//www.google-analytics.com/analytics.js","ga"),ga("create","UA-26244371-2","bedford.io"),ga("send","pageview")</script> <script src="/js/jquery.min.js"></script> <!link href="css/bootstrap.min.css" rel="stylesheet"> <script src="/js/transition.js"></script><script src="/js/collapse.js"></script><script src="/js/less.min.js"></script></head><body><div id="header"> <nav class="navbar navbar-default navbar-fixed-top" role="navigation"><div class="container"><div class="navbar-header"> <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar-collapse-element" aria-expanded="false"> <span class="sr-only">Toggle navigation</span> <span class="icon-bar"></span> <span class="icon-bar"></span> <span class="icon-bar"></span> </button> <!object class="logo navbar-brand" data="/images/logo.svg" type="image/svg+xml"><!/object> <a class="navbar-brand" href="/">The Ha Lab</a></div><div class="collapse navbar-collapse" id="navbar-collapse-element"><ul class="nav navbar-nav navbar-right"><li> <a href="/news/">News</a></li><li> <a href="/research/">Research</a></li><li> <a href="/people/">People</a></li><li> <a href="/papers/">Papers</a></li><li> <a href="/projects/">Software</a></li><li><li><a class="off" href="/join/join/">Join</a></li></li></ul></div></div> </nav></div><div class="container"><div class="hugespacer"></div><div class="bigspacer"></div><div class="row"><div class="col-md-12"><div class="title titlebox"><ol class="breadcrumb"><li><a class="off" href="/projects/ichorCNA/">ichorCNA</a></li><li>scripts</li></ol></div><p><div class="head"> Estimating tumor fraction in cell-free DNA from ultra-low-pass whole genome sequencing.</div></div></div><div class="bigspacer"></div><div class="row"><div class="col-md-3"><div class="bigspacer"></div><div class="smallhead"> Source code</div><div class="pad-left note"><div class="smallspacer"></div> <i class="fa fa-cog fa-fw"></i> <a class="off" href="https://github.com/broadinstitute/ichorCNA">github.com/broadinstitute/ichorCNA</a></div><div class="bigspacer"></div><div class="smallhead"> Contributors</div><div class="pad-left note"><div class="smallspacer"></div><div> <a class="off" href="https://github.com/gavinha"> <img class="pull-left avatar" src="https://avatars1.githubusercontent.com/u/9221688?v=4&s=50"><div class="handlebox" style="padding-left:5px""> gavinha</div> </a></div><div class="smallspacer"></div><div> <a class="off" href="https://github.com/jurhoades"> <img class="pull-left avatar" src="https://avatars3.githubusercontent.com/u/15793246?v=4&s=50"><div class="handlebox" style="padding-left:5px""> jurhoades</div> </a></div></div><div class="bigspacer"></div><div class="smallhead"> Latest commits</div><div class="pad-left smallnote"><ul class="list-unstyled"><div class="smallspacer"></div><li> <i class="fa fa-check-square-o fa-fw"></i> <a class="off" href="https://github.com/broadinstitute/ichorCNA/commit/8021afa5b72f34e8fce3c30efe0bcae9f735122d"> 10 Jul 2018 - <span class="text-gray">Added additional bin sizes for hg38 wig files #10</span> </a></li><div class="smallspacer"></div><li> <i class="fa fa-check-square-o fa-fw"></i> <a class="off" href="https://github.com/broadinstitute/ichorCNA/commit/92c8bd1e54a1221c55d590c27a91c994da62577a"> 11 May 2018 - <span class="text-gray">intersects tumour_copy with panel before normalizing, fixes #24</span> </a></li><div class="smallspacer"></div><li> <i class="fa fa-check-square-o fa-fw"></i> <a class="off" href="https://github.com/broadinstitute/ichorCNA/commit/d56208baba59256689d44dd60e60aa79e60bd14b"> 11 May 2018 - <span class="text-gray">now with mappability filter</span> </a></li><div class="smallspacer"></div><li> <i class="fa fa-check-square-o fa-fw"></i> <a class="off" href="https://github.com/broadinstitute/ichorCNA/commit/760ce1e619f262dac82f1beee3b359b0f201de5e"> 09 May 2018 - <span class="text-gray">hg38 mappability files; thanks to @leraman #22 [ci skip]</span> </a></li><div class="smallspacer"></div><li> <i class="fa fa-check-square-o fa-fw"></i> <a class="off" href="https://github.com/broadinstitute/ichorCNA/commit/05519bf02d4ae6c949b89551152cb107f86fdcb2"> 03 Apr 2018 - <span class="text-gray">fixes typo for #17</span> </a></li></ul></div><div class="spacer"></div><div class="smallhead"> Pages</div><div class="pad-left smallnote"><ul class="list-unstyled"><div class="smallspacer"></div><li> <a class="off" href="/projects/ichorCNA/"> <i class="fa fa-folder-open-o"></i> / </a></li><div class="smallspacer"></div><li> <a class="off" href="/projects/ichorCNA/scripts/snakemake/"> <i class="fa fa-folder-open-o"></i> / scripts / snakemake / </a></li><div class="smallspacer"></div><li> <a class="off" href="/projects/ichorCNA/scripts/"> <i class="fa fa-folder-open-o"></i> / scripts / </a></li></ul></div><div class="spacer"></div></div><div class="col-md-8"><div class="post"><p><h2>Contents</h2><h3>1. Snakemake Pipeline</h3><p><a href="snakemake">See instructions for snakemake pipeline</a></p><h3>2. ichorCNA R script details</h3><p>The main <a href="https://github.com/broadinstitute/ichorCNA/tree/master/scripts/scripts/runIchorCNA.R">ichorCNA R script</a> can also be run with even more exposed options once WIG files for read counts have been generated.</p><p>Here is an example of how to launch the R script from the command line: <code> Rscript ../runIchorCNA.R --libdir ../../R/ --datadir ../../inst/extdata/ \ --id tumor_sample1 --WIG results/readDepth/tumor_sample1.bin1000000.wig \ --ploidy "c(2,3)" --normal "c(0.5,0.6,0.7,0.8,0.9)" --maxCN 5 \ --includeHOMD False --chrs "c(1:22, \"X\")" --chrTrain "c(1:22)" \ --estimateNormal True --estimatePloidy True --estimateScPrevalence True \ --scStates "c(1,3)" --centromere ../../inst/extdata/GRCh37.p13_centromere_UCSC-gapTable.txt \ --exons.bed None --txnE 0.9999 --txnStrength 10000 --plotFileType png --plotYLim "c(-2,4)" \ --outDir results/ichorCNA/tumor_sample1/ &gt; logs/ichorCNA/tumor_sample1.log 2&gt; logs/ichorCNA/tumor_sample1.log </code></p><p>The list of arguments that can be passed to the script by invoking <code>--help</code>: ```</p><blockquote><p>Rscript runIchorCNA.R --help</p></blockquote><p>Usage: runIchorCNA.R [options]</p><p>Options: -l LIBDIR, --libdir=LIBDIR Script library path</p><div class="highlight"><pre><code class="language-" data-lang="">    --datadir=DATADIR
            Reference wig dir path

    -t WIG, --WIG=WIG
            Path to tumor WIG file.

    --NORMWIG=NORMWIG
            Path to normal WIG file.

    --normalPanel=NORMALPANEL
            Median corrected depth from panel of normals

    -e EXONS.BED, --exons.bed=EXONS.BED
            Path to bed file containing exon regions.

    --id=ID
            Patient ID.


    -n NORMAL, --normal=NORMAL
            Initial normal contamination

    --scStates=SCSTATES
            Subclonal states to consider

    -p PLOIDY, --ploidy=PLOIDY
            Initial tumour ploidy

    -m MAXCN, --maxCN=MAXCN
            Total clonal CN states

    --estimateNormal=ESTIMATENORMAL
            Estimate normal.

    --estimateScPrevalence=ESTIMATESCPREVALENCE
            Estimate subclonal prevalence.

    --estimatePloidy=ESTIMATEPLOIDY
            Estimate tumour ploidy.
</code></pre></div><p>(plus more arguments) ```</p></div></div><div class="col-md-1"></div></div><div class="bigspacer"></div><div class="hugespacer"></div></div><div id="footer"> <nav class="navbar navbar-default navbar-fixed-bottom" role="navigation"><div class="container"><div class="navbar-footer"> <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar-collapse-element" aria-expanded="false"> <span class="sr-only">Toggle navigation</span> <span class="icon-bar"></span> <span class="icon-bar"></span> <span class="icon-bar"></span> </button></div> <!object class="logofooter navbar-brand" data="/images/fhcrc-logo.png" type="image/png"><!/object><div class="collapse navbar-collapse" id="navbar-collapse-element"><ul class="nav navbar-nav navbar-right navbar-tiny"><li><a class="off" href="/misc/about/">About</a></li><li><a class="off" href="/misc/contact/">Contact</a></li></ul><ul class="nav navbar-text"></ul></div></div> </nav></div></body></html>