---
layout: misc
course: Genome 541
name: Introduction to Computational Molecular Biology
title: Genome 541 - Introduction to Computational Molecular Biology
prof: Gavin Ha
location: UW
year: 2020
semester: Spring
website: https://noble.gs.washington.edu/~wnoble/genome541/
---

## Cancer Genomics Module

#### [Genome 541 Course Website](https://noble.gs.washington.edu/~wnoble/genome541/){:target="_blank"}

<!-- #### ***IMPORTANT: Due to the policies enacted by UW in response to the COVID-19 health crisis, lectures will be instructed online only.*** -->

The Cancer Genomics Module is the third module in the course and consists of 4 lectures. <br>
Homework #5 and #6 accompany this module.


### Schedule of lectures and homework
**Dates:** April 28 - May 7

**Times:** Tuesday & Thursday @ 10:30am - 11:50am

**TA:** Anna-Lisa Doebley (adoebley@uw.edu)

**Office Hours (Virtual):** 2-3pm on Mon 5/4, Wed 5/6, Mon 5/11, Wed 5/13 <br>
Zoom Meeting ID: [446 356 7725](https://washington.zoom.us/j/4463567725){:target="_blank"} (password protected)

<style>
      table, td, th { 
      padding: 10px; 
      border: 2px solid #1c87c9;
      border-radius: 5px;
      background-color: #ffffff;
      text-align: center;
      }
    </style>
<table>
	<tr>
		<th width="10%">Date</th>
		<th width="50%" style="text-align:center">Lecture Title</th>
		<th width="15" style="text-align:center">Lecture Slides</th>
		<th width="25%">Homework Assigned</th>
	</tr>
	<tr>
		<td>April 28</td>
		<td><a href="#lecture-1-introduction-to-cancer-genome-analysis">Introduction to Cancer Genome Analysis</a></td>
		<td><a href="./2020/GS541_CancerGenomics_Lecture1.pdf" target="_blank">Lecture 1</a></td>
		<td><a href="#homework">Homework 5</a></td>
	</tr>
	<tr>
		<td>April 30</td>
		<td><a href="#lecture-2-probabilistic-methods-for-mutation-detection">Probabilistic Methods for Mutation Detection</a></td>
		<td><a href="./2020/GS541_CancerGenomics_Lecture2.pdf" target="_blank">Lecture 2</a></td>
		<td></td>
	</tr>
	<tr>
		<td>May 5</td>
		<td><a href="#lecture-3-probabilistic-methods-for-copy-number-alteration-detection">Probabilistic Methods for Copy Number Alteration Detection</a></td>
		<td><a href="./2020/GS541_CancerGenomics_Lecture3.pdf" target="_blank">Lecture 3</a></td>
		<td><a href="#homework">Homework 6</a></td>
	</tr>
	<tr>
		<td>May 7</td>
		<td><a href="#lecture-4-additional-topics">Additional Topics</a></td>
		<td></td>
		<td></td>
	</tr>
</table>

<br>

#### Homework
<table>
	<tr>
		<th width="30%">Homework</th>
		<th width="50%">Files</th>
		<th width="20%">Due Date</th>
	</tr>
	<tr>
		<td>Homework #5:<br>Single nucleotide variant genotyping</td>
		<td style="text-align:left"><a href="./2020/Homework5/Homework5_SNVGenotyping_Assignment.pdf" target="_blank">1. Assignment</a><br>
			<a href="./2020/Homework5/Homework5_SNVGenotyping_R-template.Rmd" target="_blank">2. R Markdown template</a><br>
			<a href="./2020/Homework5/Homework5_SNVGenotyping_python-template.ipynb" target="_blank">3. Python Jupyter notebook template</a><br>
			<a href="./2020/Homework5/Homework5_alleleCounts.txt" target="_blank">4. Homework5_alleleCounts.txt</a></td>
		<td>May 8, 11:59pm</td>
	</tr>
	<tr>
		<td>Homework #6:<br>Profiling copy number alterations</td>
		<td style="text-align:left"></td>
		<td>May 15, 11:59pm</td>
	</tr>
</table>

<br><br>

### Module Outline

### Lecture 1: Introduction to cancer genome analysis
[Lecture 1 Slides](./2020/GS541_CancerGenomics_Lecture1.pdf){:target="_blank"}

1. Background on Cancer Genome Alterations
	- Genomic alterations in cancer: drivers vs passengers, somatic vs germline
	- Tumor evolution and heterogeneity
2. Overview of Cancer Genome Analysis
	- Computational strategy and workflow
	- Tumor DNA Sequencing 
	- Types of genomic alterations predicted from tumor sequencing
	- Methods/tools/algorithms in following lectures
3. Primer on statistical modeling  
	- Probability distribution, Bayesian statistics, inference


<br><br>
### Lecture 2: Probabilistic methods for mutation detection
[Lecture 2 Slides](./2020/GS541_CancerGenomics_Lecture2.pdf){:target="_blank"}

1. Primer on statistical modeling (contâ€™d)
	- Mixture models and inference using the EM algorithm
2. Detecting Mutations in Cancer Genomes
	- Visualizing somatic vs germline SNVs
	- Sequencing read count data
	- SNV genotyping strategy
3. Mixture Models for SNV Detection
	- SNVMix probabilistic model and EM inference
	- Predicting somatic SNVs in cancer

- References: <br>
[SNVMix: Goya et al. Bioinformatics (2010)](./2020/Papers/Goya2010_SNVMix.pdf){:target="_blank"} <br>
[JointSNVMix: Roth et al. Bioinformatics (2012)](./2020/Papers/Roth2012_JointSNVMix.pdf){:target="_blank"}

<br><br>
### Lecture 3: Probabilistic methods for copy number alteration detection
[Lecture 3 Slides](./2020/GS541_CancerGenomics_Lecture3.pdf){:target="_blank"}

1. Detecting Copy Number Alterations in Cancer Genomes
	- Predicting copy number features from sequence data 
	- Copy number analysis workflow
	- Data normalization 
2. Continuous Hidden Markov Model (HMM)
	- Graphical model representation
	- Components of a continuous HMM
	- Inference & parameter estimation using expectation-maximization (EM)
3. Copy Number Profiling using a Hidden Markov Model
	- Probabilistic model for copy number analysis
	- Predicting copy number segments using the Viterbi algorithm

- References: <br>
[HMMcopy: Ha et al. Genome Res (2012)](./2020/Papers/Ha2012_HMMcopy.pdf){:target="_blank"} <br>
[ichorCNA: Adalsteisson et al. Nat Commun (2017)](./2020/Papers/Adalsteinsson2017_ichorCNA.pdf){:target="_blank"}

<br><br>
### Lecture 4: Additional topics
[Slides]()

1. Estimating Intra-Tumor Heterogeneity 
	- ichorCNA: Modeling tumor-normal admixture
	- Modeling tumor clonality 
2. Assessing Statistical Power for Variant Discovery
	- Power calculation
	- Calibrating sequencing depth for variant discovery

- References: <br>
[ichorCNA: Adalsteisson et al. Nat Commun (2017)](./2020/Papers/Adalsteinsson2017_ichorCNA.pdf){:target="_blank"} <br>
[TitanCNA: Ha et al. Genome Res (2014)](./2020/Papers/Ha2014_TITAN.pdf){:target="_blank"}




