---
layout: misc
title: Research
---

## Overview
The Ha laboratory is interested in studying the role of genomic alterations in cancer progression and translating this knowledge to expand applications for precision medicine. 

We combine research in *two* complementary areas:  

1.	Develop and apply novel **computational methods** to comprehensively profile **cancer genomes** from tumor tissue in large patient cohorts. 
2.	Develop approaches to exploit **liquid biopsies**, such as circulating **cell-free DNA** from blood plasma, for studying cancer. 

We leverage insights from the analysis of tumor genomes to inform the design of blood-based applications to monitor patient response to treatment. Our goals are to uncover mechanisms of treatment resistance, to identify blood-based genetic biomarkers, and to translate these findings to help improve clinical decisions. We are actively engaged in collaborations with research and clinical colleagues, including at [Fred Hutchinson Cancer Center](https://www.fredhutch.org/en.html){:target="_blank"}, [UW Medicine](https://www.uwmedicine.org/){:target="_blank"}, and [Dana-Farber Cancer Institute](https://www.dana-farber.org/){:target="_blank"} to study different tumor types.


<figure style="float: right;">
	<img src="/images/research/cfDNA_workflow-01.png" alt="Whole genome and exome analysis of cell-free DNA from cancer patients." width="250"/>
	<figcaption><center><b>Fig. </b>Analysis of cell-free DNA<br>from cancer patients.</center></figcaption>
</figure>

### 1. Liquid Biopsies to Study Cancer
Liquid biopsies, such as blood plasma containing cell-free DNA (cfDNA), provide a non-invasive route to study cancer. In cancer patients, cfDNA can contain circulating tumor DNA (ctDNA) released from dying tumor cells. ctDNA is more accessible than tumor tissue, particularly in advanced stage cancers when tumor biopsies are difficult to obtain. We are interested in the framework of using ctDNA for enhancing the study of large cancer patient cohorts that have less accessible tumors. The strategy of studying both tumor and cfDNA sequencing will help to establish a new paradigm in large-scale cancer genome studies.

#### <em>Estimating ctDNA content using cost-effective approach</em>
Our lab develops novel computational approaches and pipelines for high-sensitivity copy number and tissue-of-origin analysis from the sequencing of cfDNA. We developed the tool, **[ichorCNA](https://github.com/GavinHaLab/ichorCNA){:target="_blank"}**, which estimates the tumor fraction and copy number alterations from cost-effective ultra-low pass genome sequencing ([Adalsteinsson\*, Ha\*, Freeman\* et al. Nat Commun, 2017](/papers/2-adalsteinsson-ha-2017-natcomms-cfdna/)). 

#### <em>Tumor phenotype analysis from cell-free DNA</em>
Current research and clinical efforts have focused on mutation detection in select cancer genes from ctDNA. However, approaches to characterize the tumor phenotype from ctDNA is still a nascent area of research and development.

Recently, we developed **[Griffin](https://github.com/GavinHaLab/Griffin){:target="_blank"}**, which is a framework for profiling nucleosome accessibility from cell-free DNA for studying transcriptional regulation and tumor phenotypes ([Doebley et al. medRxiv, 2021](/papers/doebley_medrixv/)). We have applied these to classify breast cancer hormone status ([Doebley et al. medRxiv, 2021](/papers/doebley_medrixv/)) 
To study the transcriptional activity from ctDNA, we developed **[TritonNP](https://github.com/GavinHaLab/TritonNP){:target="_blank"}** to capture nucleosome patterns associated with active transcription in key regulators of prostate cancer ([De Sarkar* and Patton* et al. bioRxiv, 2022](/papers/desarkar-biorxiv/)).

<figure style="float: left;">
	<img src="/images/research/Griffin_schematic-01.png" alt="Brief Griffin schematic." width="210"/>
	<figcaption><center><b>Fig. </b>Griffin framework for <br>cancer detection and tumor subtyping.<br>
	(<a href="/papers/desarkar-biorxiv/" target="_blank">Doebley et al. <em>medRxiv</em>, 2021</a>; Fig 1)</center></figcaption>
</figure>

<figure style="float: center;">
	<center><img src="/images/research/PDX_schematic-01.png" alt="PDX mouse plasma." width="450"/></center>
	<figcaption><center><b>Fig. </b>Plasma ctDNA study from patient derived xenografts.<br>
	(<a href="/papers/desarkar-biorxiv/" target="_blank">De Sarkar*, Patton* et al. <em>bioRxiv</em>, 2022</a>; Fig 1A and 3A)</center></figcaption>
</figure>

<br><br><br><br>

<figure style="float: right;">
	<!-- <img src="/images/research/TF_PSA_track.png" alt="Circulating tumor DNA fraction tracks with prostate specific antigen." width="325"/>
	<figcaption><center><b>Fig. </b>Circulating tumor DNA fraction<br>tracks with PSA biomarker.</center></figcaption> -->
	<center><img src="/images/research/AR_enh_CN_cfDNA_track.png" alt="AR gene and enhancer copy number tracks with PSA." width="200"/></center>
	<figcaption><center><b>Fig. </b> Copy number of driver genes track with PSA.<br>(<a href="/papers/4-viswanathan-ha-2018-cell-crpc-10x/" target="_blank">Viswanathan*, Ha*, Hoff* et al. <em>Cell</em>, 2018</a>; Fig 7C)</center></figcaption>
	<br>
	<center><img src="/images/research/Patient_phenotyping-01.png" alt="Patient phenotyping." width="200"/></center>
	<figcaption><center><b>Fig. </b>Methods for prostate tumor phenotype <br>prediction in patient plasma samples.<br>
	(<a href="/papers/desarkar-biorxiv/" target="_blank">De Sarkar*, Patton* et al. <em>bioRxiv</em>, 2022</a>; Fig 1)</center></figcaption>
</figure>

### 2. Development of Methods for Precision Oncology
A challenge in precision medicine is the ability to routinely sequence patients’ tumors because repeated biopsies are not feasible, especially for advanced cancers. Liquid biopsy is an attractive solution for monitoring tumor genomic alterations from circulating tumor DNA (ctDNA). However, the systematic analysis to detect genome-wide signatures from ctDNA is still needed to understand the feasibility of longitudinally monitoring the dynamics of tumor clones in circulation. The lack of robust informatics solutions is a major barrier to realizing the potential of cfDNA. The development of sensitive approaches for analyzing ctDNA are needed to further strengthen the utility of cfDNA for cancer genome diagnostics, identification of biomarkers, and patient monitoring of response to therapy. 

#### <em>Monitoring genomic alterations from blood during therapy</em>
We are developing novel algorithms to detect and validate tumor-specific alterations in ctDNA and to use these aberrant genomic events as reliably markers to monitor patient response to treatment. By increasing the sensitivity for detecting alterations in ctDNA, we will work towards important applications for clinical tracking of cancer burden in patients, perhaps even early stage cancer detection. 

#### <em>Classifying tumor subtypes and phenotypes in patient plasma</em>
phenotypes of adenocarinoma and neuroendocrine tumors (using [Keraon](https://github.com/GavinHaLab/Keraon){:target="_blank"} [ctdPheno](https://github.com/GavinHaLab/CRPCSubtypingPaper/tree/main/ctdPheno){:target="_blank"} 
[De Sarkar* and Patton* et al. bioRxiv, 2022](/papers/desarkar-biorxiv/)).

<br><br><br><br><br>


<figure style="float: right;">
	<img src="/images/research/AR_SV.png" alt="Structural rearrangements in cancer genomes." width="400"/>
	<figcaption><center><b>Fig.</b> Structural rearrangements in cancer genomes<br>(<a href="/papers/4-viswanathan-ha-2018-cell-crpc-10x/" target="_blank">Viswanathan*, Ha*, Hoff* et al. <em>Cell</em>, 2018</a>; Fig 4A).</center></figcaption>
</figure>

### 3. Application of New Technologies to Study Genomic Alterations in Tumor Genomes 
Analysis of whole genome sequencing (WGS) of tumor genomes can uncover genome-wide alterations, including structural rearrangements and mutations in non-coding features, which are not easily interrogated by whole exome sequencing (WES). Genomic rearrangements can alter non-coding features, including the duplication of enhancers of oncogenes and relocation (‘hijacking’) of enhancers to proto-oncogenes. The reconstruction of these genome rearrangements will be important to understanding the their role in cancer.

Our lab is interested in studying the abnormal structure of the cancer genome more deeply. We apply cutting-edge whole genome DNA sequencing technologies, particularly platforms that generate long-range genomic information such as linked-read and long-read data. These data enhance the reconstruction of genomic rearrangements and enable the study of alterations in non-coding genomic regions. We are also interested in integrating chromosome conformation information to better understand the effects of genomic alterations on the 3D chromosome structure. 

Our work in advanced prostate and breast cancers, as well as work by others, have revealed non-coding structural rearrangements altering enhancer elements and driving expression of cancer genes. There is a need for more comprehensive analyses of cancer genomes to uncover new non-coding alterations driving cancer progression. We are interesting in performing whole genome characterization of metastatic cancers, from both tumor and liquid biopsies, to address major research questions, such as treatment resistance.

<br>

### 4. Development of Computational Methods for Cancer Genomics
<figure style="float: right;">
	<img src="/images/research/Titan_haplotype_graphicalmodel.png" alt="Probabilistic graphical model of TITAN" width="200"/>
	<figcaption><center><b>Fig.</b> Probabilistic model of TITAN<br>(<a href="/papers/1-ha-2014-gr-titan/" target="_blank">Ha et al. <em>Genome Res</em>, 2014</a>)</center></figcaption>
</figure>
We develop tools that employ machine learning algorithms to analyze cancer genomes. We use statistical/probabilistic models to develop novel algorithms and frameworks for predicting the aberrant structure of rearranged genomes in cancer. These include tools such as **[TITAN](https://github.com/gavinha/TitanCNA){:target="_blank"}**, which predicts subclonal copy number in tumors with intra-tumor heterogeneity.

Current genome sequencing approaches have limited capabilities for interrogating low complexity, repetitive non-coding regions and for accurate reconstruction of complex rearrangements. The advent of new experimental technologies to generate long-range genomic information, such as linked-read sequencing (10X Genomics), is beginning to address these deficiencies. However, these approaches are still in their infancy and lack robust analytical tools to decipher these data and to realize their potential for understanding cancer genome structure. We have extended or modified current algorithms to leverage these new data features to improve prediction of copy number alterations and structural rearrangements (see **[Software](/projects/)**).

<br><br><br><br>

# Funding
The Gavin Ha Lab is generously supported by funding from the following institutes, foundations, and agencies. Dr. Ha is also supported by the following awards:
- **NIH Director's New Innovative Award** ([DP2 CA280624](https://reporter.nih.gov/search/STFly4W4Rk6mvc7irU8hbw/project-details/10473384){:target="_blank"})
- **Transition Career Development Award** (NCI, [K22 CA237746](https://projectreporter.nih.gov/project_info_description.cfm?aid=9720173&icde=50210569){:target="_blank"}), 
- **Prostate Cancer Foundation Young Investigator Award** ([PCF YIA](https://www.pcf.org/c/young-investigator-award-class-of-2019/#Ha){:target="_blank"})
- [**V Scholar Grant**](https://www.v.org/grants-awarded/to/gavin-ha-ph-d/){:target="_blank"} from The V Foundation

<figure style="float: center;">
	<img src="/images/research/FH_logo.png" alt="Fred Hutch" width="275"/>
	<img src="/images/research/BBI_logo.png" alt="Brotman-Baty Institute Catalytic Collaboration Grant" width="275"/>
	<img src="/images/research/ccsg_logo-01.png" alt="Cancer Consortium Support Grant" width="275"/>
	<img src="/images/research/v_foundation_logo1.png" alt="V Foundation V Scholar Grant" width="180"/>
	<img src="/images/research/pcf_logo.png" alt="Prostate Cancer Foundation Young Investigator Award" width="250"/>
	<img src="/images/research/nci_logo.png" alt="NCI K22 Career Transition Grant" width="300"/>	
	<img src="/images/research/ici_logo.png" alt="Fund for Innovation in Cancer Informatics Major Grant" width="85"/>
	<img src="/images/research/DOD_logo.jpeg" alt="CDMRP/DoD Prostate Cancer Research Program" width="220"/>
	<img src="/images/research/kuni.jpeg" alt="Kuni Foundation Cancer Discovery Grant" width="170"/>
</figure>




