<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"/><title>TitanCNA/scripts/R_scripts/</title><meta name="viewport" content="width=device-width, initial-scale=1.0"><link rel="alternate" type="application/rss+xml" title="Ha Lab" href="http://gavinha.github.io/feed.xml"><link href="/css/style.css" rel="stylesheet" media="screen"> <script>!function(e){var t,a={kitId:"xeu8jut",scriptTimeout:3e3},c=e.documentElement,i=setTimeout(function(){c.className=c.className.replace(/\bwf-loading\b/g,"")+" wf-inactive"},a.scriptTimeout),n=e.createElement("script"),o=!1,s=e.getElementsByTagName("script")[0];c.className+=" wf-loading",n.src="//use.typekit.net/"+a.kitId+".js",n.async=!0,n.onload=n.onreadystatechange=function(){if(t=this.readyState,!(o||t&&"complete"!=t&&"loaded"!=t)){o=!0,clearTimeout(i);try{Typekit.load(a)}catch(e){}}},s.parentNode.insertBefore(n,s)}(document)</script> <script>!function(e,a,n,t,c,o,s){e.GoogleAnalyticsObject=c,e[c]=e[c]||function(){(e[c].q=e[c].q||[]).push(arguments)},e[c].l=1*new Date,o=a.createElement(n),s=a.getElementsByTagName(n)[0],o.async=1,o.src=t,s.parentNode.insertBefore(o,s)}(window,document,"script","//www.google-analytics.com/analytics.js","ga"),ga("create","UA-26244371-2","bedford.io"),ga("send","pageview")</script> <script src="/js/jquery.min.js"></script> <!link href="css/bootstrap.min.css" rel="stylesheet"> <script src="/js/transition.js"></script><script src="/js/collapse.js"></script><script src="/js/less.min.js"></script></head><body><div id="header"> <nav class="navbar navbar-default navbar-fixed-top" role="navigation"><div class="container"><div class="navbar-header"> <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar-collapse-element" aria-expanded="false"> <span class="sr-only">Toggle navigation</span> <span class="icon-bar"></span> <span class="icon-bar"></span> <span class="icon-bar"></span> </button> <!object class="logo navbar-brand" data="/images/logo.svg" type="image/svg+xml"><!/object> <a class="navbar-brand" href="/">Ha Lab</a></div><div class="collapse navbar-collapse" id="navbar-collapse-element"><ul class="nav navbar-nav navbar-right"><li> <a href="/news/">News</a></li><li> <a href="/research/">Research</a></li><li> <a href="/people/">People</a></li><li> <a href="/papers/">Papers</a></li><li> <a href="/projects/">Software</a></li><li><li><a class="off" href="/join/join/">Join</a></li></li></ul></div></div> </nav></div><div class="container"><div class="hugespacer"></div><div class="bigspacer"></div><div class="row"><div class="col-md-12"><div class="title titlebox"><ol class="breadcrumb"><li><a class="off" href="/projects/TitanCNA/">TitanCNA</a></li><li><a class="off" href="/projects/TitanCNA/scripts/">scripts</a></li><li>R_scripts</li></ol></div><p><div class="head"> Analysis of subclonal copy number alterations (CNA) and loss of heterozygosity (LOH) in cancer</div></div></div><div class="bigspacer"></div><div class="row"><div class="col-md-3"><div class="bigspacer"></div><div class="smallhead"> Source code</div><div class="pad-left note"><div class="smallspacer"></div> <i class="fa fa-cog fa-fw"></i> <a class="off" href="https://github.com/gavinha/TitanCNA">github.com/gavinha/TitanCNA</a></div><div class="bigspacer"></div><div class="smallhead"> Contributors</div><div class="pad-left note"><div class="smallspacer"></div><div> <a class="off" href="https://github.com/gavinha"> <img class="pull-left avatar" src="https://avatars1.githubusercontent.com/u/9221688?v=4&s=50"><div class="handlebox" style="padding-left:5px""> gavinha</div> </a></div><div class="smallspacer"></div><div> <a class="off" href="https://github.com/lbeltrame"> <img class="pull-left avatar" src="https://avatars0.githubusercontent.com/u/25192?v=4&s=50"><div class="handlebox" style="padding-left:5px""> lbeltrame</div> </a></div><div class="smallspacer"></div><div> <a class="off" href="https://github.com/dtenenba"> <img class="pull-left avatar" src="https://avatars0.githubusercontent.com/u/2286826?v=4&s=50"><div class="handlebox" style="padding-left:5px""> dtenenba</div> </a></div><div class="smallspacer"></div><div> <a class="off" href="https://github.com/hpages"> <img class="pull-left avatar" src="https://avatars1.githubusercontent.com/u/8810451?v=4&s=50"><div class="handlebox" style="padding-left:5px""> hpages</div> </a></div><div class="smallspacer"></div><div> <a class="off" href="https://github.com/mrjc42"> <img class="pull-left avatar" src="https://avatars3.githubusercontent.com/u/2029097?v=4&s=50"><div class="handlebox" style="padding-left:5px""> mrjc42</div> </a></div></div><div class="bigspacer"></div><div class="smallhead"> Latest commits</div><div class="pad-left smallnote"><ul class="list-unstyled"><div class="smallspacer"></div><li> <i class="fa fa-check-square-o fa-fw"></i> <a class="off" href="https://github.com/gavinha/TitanCNA/commit/e46683c75d28961668d10882294ea0ab53975685"> 26 Jun 2018 - <span class="text-gray">fixed bug in getSubcloneProfiles when sample has no CNA or too noisy #22</span> </a></li><div class="smallspacer"></div><li> <i class="fa fa-check-square-o fa-fw"></i> <a class="off" href="https://github.com/gavinha/TitanCNA/commit/d2687ea377ba2c59005b66fca5fc4fd6a9dec4c6"> 24 May 2018 - <span class="text-gray">fixed bug in selectSolution.R [ci skip]</span> </a></li><div class="smallspacer"></div><li> <i class="fa fa-check-square-o fa-fw"></i> <a class="off" href="https://github.com/gavinha/TitanCNA/commit/1b0d954293435c0fdd51fce769eba01b3164928c"> 11 Apr 2018 - <span class="text-gray">version bump v1.17.2</span> </a></li><div class="smallspacer"></div><li> <i class="fa fa-check-square-o fa-fw"></i> <a class="off" href="https://github.com/gavinha/TitanCNA/commit/7fc4fcc3192aa975a480ed360db0db36d46a6ae6"> 11 Apr 2018 - <span class="text-gray">include function getHaplotypesFromVCF [ci skip]</span> </a></li><div class="smallspacer"></div><li> <i class="fa fa-check-square-o fa-fw"></i> <a class="off" href="https://github.com/gavinha/TitanCNA/commit/635e89acd2fa1b75adca3c7438aa71c8ef0e8586"> 05 Apr 2018 - <span class="text-gray">remove redundant ichorCNA options [ci skip]</span> </a></li></ul></div><div class="spacer"></div><div class="smallhead"> Pages</div><div class="pad-left smallnote"><ul class="list-unstyled"><div class="smallspacer"></div><li> <a class="off" href="/projects/TitanCNA/scripts/R_scripts/"> <i class="fa fa-folder-open-o"></i> / scripts / R_scripts / </a></li><div class="smallspacer"></div><li> <a class="off" href="/projects/TitanCNA/scripts/snakemake/"> <i class="fa fa-folder-open-o"></i> / scripts / snakemake / </a></li><div class="smallspacer"></div><li> <a class="off" href="/projects/TitanCNA/scripts/TenX_scripts/"> <i class="fa fa-folder-open-o"></i> / scripts / TenX_scripts / </a></li><div class="smallspacer"></div><li> <a class="off" href="/projects/TitanCNA/scripts/"> <i class="fa fa-folder-open-o"></i> / scripts / </a></li><div class="smallspacer"></div><li> <a class="off" href="/projects/TitanCNA/"> <i class="fa fa-folder-open-o"></i> / </a></li></ul></div><div class="spacer"></div></div><div class="col-md-8"><div class="post"><p><h1 id="r-scripts-for-titancna-analysis">R scripts for TitanCNA analysis</h1><h2 id="standard-whole-genomeexome-sequencing-analysis">1. Standard Whole Genome/Exome Sequencing Analysis</h2><p>R script (<code class="highlighter-rouge">titanCNA.R</code>) for running TitanCNA analysis on standard whole genome and exome sequencing data.</p><p><strong>Input files</strong><br/> This script assumes that the necessary input files have been generated. These are generated by the KRONOS workflow.</p><ol><li>GC-corrected, normalized read coverage using the HMMcopy suite<br/> * For exome analysis, please use the <code class="highlighter-rouge">targetedSequence</code> argument to specify the dataframe containing the exon baits input from a bed file.</li></ol><div class="highlighter-rouge"><pre class="highlight"><code>exons &lt;- read.delim("exon_baits.bed", header = TRUE, as.is = TRUE)
correctReadDepth(tumWig, normWig, gcWig, mapWig, genomeStyle = "NCBI", targetedSequence = exons)
</code></pre></div><ol><li>Tumour allelic read counts at heterozygous SNPs (identifed from the normal sample).</li></ol><p><strong>Running the R script</strong></p><ol><li>Look at the usage of the R script<br/> ``` # from the command line<blockquote><p>Rscript titanCNA.R –help Usage: Rscript titanCNA.R [options]</p></blockquote></li></ol><p>Options: –id=ID Sample ID</p><div class="highlighter-rouge"><pre class="highlight"><code>    --hetFile=HETFILE
            File containing allelic read counts at HET sites. (Required)

    --cnFile=CNFILE
            File containing normalized coverage as log2 ratios. (Required)

    --outDir=OUTDIR
            Output directory to output the results. (Required)

    --numClusters=NUMCLUSTERS
            Number of clonal clusters. (Default: 1)

    --numCores=NUMCORES
            Number of cores to use. (Default: 1)

    --ploidy_0=PLOIDY_0
            Initial ploidy value; float (Default: 2)

    --estimatePloidy=ESTIMATEPLOIDY
            Estimate ploidy; TRUE or FALSE (Default: TRUE)

    --normal_0=NORMAL_0
            Initial normal contamination (1-purity); float (Default: 0.5)

    --estimateNormal=ESTIMATENORMAL
            Estimate normal contamination method; string {'map', 'fixed'} (Default: map)

    --maxCN=MAXCN
            Maximum number of copies to model; integer (Default: 8)
    (... additional arguments)   ```
</code></pre></div><p>Additional arguments to consider are the following:<br/> These arguments can be used to tune the model based on variance in the read coverage data and data-type (whole-exome sequencing or whole-genome sequencing). ``` –alphaK=ALPHAK Hyperparameter on Gaussian variance; for WES, use 2500; for WGS, use 10000; float (Default: 10000)</p><p>–alphaKHigh=ALPHAKHIGH Hyperparameter on Gaussian variance for extreme copy number states; for WES, use 2500; for WGS, use 10000; float (Default: 10000) ```</p><ol><li><p>Example usage of R script <code class="highlighter-rouge"> # normalized coverage file: test.cn.txt # allelic read count file: test.het.txt Rscript titanCNA.R --id test --hetFile test.het.txt --cnFile test.cn.txt \ --numClusters 1 --numCores 1 --normal_0 0.5 --ploidy_0 2 \ --chrs "c(1:22, \"X\")" --estimatePloidy TRUE --outDir ./ </code></p></li><li><p>Running TitanCNA for multiple restarts and model selection <code class="highlighter-rouge">titanCNA.R</code> should be run with multiple restarts for different values of (a) Ploidy (2,3,4) and (b) Number of clonal clusters. This will lead to multiple solutions. Each set of solutions for a given initialization of ploidy value will be saved to a directory (e.g. run_ploidy2, run_ploidy3, run_ploidy4). The R script <code class="highlighter-rouge">selectSolution.R</code> will help select the optimal cluster from all these solutions. The output is a tab-delimited file indicating the selected solution, along with parameters for that run. It also includes the path to the results so users can collect the results. ``` numClusters=3 numCores=4 ## run TITAN for each ploidy (2,3,4) and clusters (1 to numClusters) echo “Maximum number of clusters: $numClusters”; for ploidy in $(seq 2 4) do echo “Running TITAN for $i clusters.”; outDir=run_ploidy$ploidy mkdir $outDir for numClust in $(seq 1 $numClusters) do echo “Running for ploidy=$ploidy”; Rscript titanCNA_v1.10.1.R –id test –hetFile test.het.txt –cnFile test.cn.txt \ –numClusters $numClust –numCores $numCores –normal_0 0.5 –ploidy_0 $ploidy \ –chrs “c(1:22, "X")” –estimatePloidy TRUE –outDir $outDir done echo “Completed job for $numClust clusters.” done</p></li></ol><p>## select optimal solution Rscript selectSolution.R –ploidyRun2=run_ploidy2 –ploidyRun3=run_ploidy3 –ploidyRun4=run_ploidy4 –threshold=0.05 –outFile optimalClusters.txt ```</p></div></div><div class="col-md-1"></div></div><div class="bigspacer"></div><div class="hugespacer"></div></div><div id="footer"> <nav class="navbar navbar-default navbar-fixed-bottom" role="navigation"><div class="container"><div class="navbar-footer"> <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar-collapse-element" aria-expanded="false"> <span class="sr-only">Toggle navigation</span> <span class="icon-bar"></span> <span class="icon-bar"></span> <span class="icon-bar"></span> </button></div> <!object class="logofooter navbar-brand" data="/images/fhcrc-logo.png" type="image/png"><!/object><div class="collapse navbar-collapse" id="navbar-collapse-element"><ul class="nav navbar-nav navbar-right navbar-tiny"><li><a class="off" href="/misc/about/">About</a></li><li><a class="off" href="/misc/contact/">Contact</a></li></ul><ul class="nav navbar-text"></ul></div></div> </nav></div></body></html>