<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"/><title>TitanCNA/scripts/TenX_scripts/</title><meta name="viewport" content="width=device-width, initial-scale=1.0"><link rel="alternate" type="application/rss+xml" title="Ha Lab" href="http://gavinha.github.io/feed.xml"><link href="/css/style.css" rel="stylesheet" media="screen"> <script>!function(e){var t,a={kitId:"xeu8jut",scriptTimeout:3e3},c=e.documentElement,i=setTimeout(function(){c.className=c.className.replace(/\bwf-loading\b/g,"")+" wf-inactive"},a.scriptTimeout),n=e.createElement("script"),o=!1,s=e.getElementsByTagName("script")[0];c.className+=" wf-loading",n.src="//use.typekit.net/"+a.kitId+".js",n.async=!0,n.onload=n.onreadystatechange=function(){if(t=this.readyState,!(o||t&&"complete"!=t&&"loaded"!=t)){o=!0,clearTimeout(i);try{Typekit.load(a)}catch(e){}}},s.parentNode.insertBefore(n,s)}(document)</script> <script>!function(e,a,n,t,c,o,s){e.GoogleAnalyticsObject=c,e[c]=e[c]||function(){(e[c].q=e[c].q||[]).push(arguments)},e[c].l=1*new Date,o=a.createElement(n),s=a.getElementsByTagName(n)[0],o.async=1,o.src=t,s.parentNode.insertBefore(o,s)}(window,document,"script","//www.google-analytics.com/analytics.js","ga"),ga("create","UA-26244371-2","bedford.io"),ga("send","pageview")</script> <script src="/js/jquery.min.js"></script> <!link href="css/bootstrap.min.css" rel="stylesheet"> <script src="/js/transition.js"></script><script src="/js/collapse.js"></script><script src="/js/less.min.js"></script></head><body><div id="header"> <nav class="navbar navbar-default navbar-fixed-top" role="navigation"><div class="container"><div class="navbar-header"> <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar-collapse-element" aria-expanded="false"> <span class="sr-only">Toggle navigation</span> <span class="icon-bar"></span> <span class="icon-bar"></span> <span class="icon-bar"></span> </button> <!object class="logo navbar-brand" data="/images/logo.svg" type="image/svg+xml"><!/object> <a class="navbar-brand" href="/">Ha Lab</a></div><div class="collapse navbar-collapse" id="navbar-collapse-element"><ul class="nav navbar-nav navbar-right"><li> <a href="/news/">News</a></li><li> <a href="/research/">Research</a></li><li> <a href="/people/">People</a></li><li> <a href="/papers/">Papers</a></li><li> <a href="/projects/">Software</a></li><li><li><a class="off" href="/join/join/">Join</a></li></li></ul></div></div> </nav></div><div class="container"><div class="hugespacer"></div><div class="bigspacer"></div><div class="row"><div class="col-md-12"><div class="title titlebox"><ol class="breadcrumb"><li><a class="off" href="/projects/TitanCNA/">TitanCNA</a></li><li><a class="off" href="/projects/TitanCNA/scripts/">scripts</a></li><li>TenX_scripts</li></ol></div><p><div class="head"> Analysis of subclonal copy number alterations (CNA) and loss of heterozygosity (LOH) in cancer</div></div></div><div class="bigspacer"></div><div class="row"><div class="col-md-3"><div class="bigspacer"></div><div class="smallhead"> Source code</div><div class="pad-left note"><div class="smallspacer"></div> <i class="fa fa-cog fa-fw"></i> <a class="off" href="https://github.com/gavinha/TitanCNA">github.com/gavinha/TitanCNA</a></div><div class="bigspacer"></div><div class="smallhead"> Contributors</div><div class="pad-left note"><div class="smallspacer"></div><div> <a class="off" href="https://github.com/gavinha"> <img class="pull-left avatar" src="https://avatars1.githubusercontent.com/u/9221688?v=4&s=50"><div class="handlebox" style="padding-left:5px""> gavinha</div> </a></div><div class="smallspacer"></div><div> <a class="off" href="https://github.com/lbeltrame"> <img class="pull-left avatar" src="https://avatars0.githubusercontent.com/u/25192?v=4&s=50"><div class="handlebox" style="padding-left:5px""> lbeltrame</div> </a></div><div class="smallspacer"></div><div> <a class="off" href="https://github.com/dtenenba"> <img class="pull-left avatar" src="https://avatars0.githubusercontent.com/u/2286826?v=4&s=50"><div class="handlebox" style="padding-left:5px""> dtenenba</div> </a></div><div class="smallspacer"></div><div> <a class="off" href="https://github.com/hpages"> <img class="pull-left avatar" src="https://avatars1.githubusercontent.com/u/8810451?v=4&s=50"><div class="handlebox" style="padding-left:5px""> hpages</div> </a></div><div class="smallspacer"></div><div> <a class="off" href="https://github.com/mrjc42"> <img class="pull-left avatar" src="https://avatars3.githubusercontent.com/u/2029097?v=4&s=50"><div class="handlebox" style="padding-left:5px""> mrjc42</div> </a></div></div><div class="bigspacer"></div><div class="smallhead"> Latest commits</div><div class="pad-left smallnote"><ul class="list-unstyled"><div class="smallspacer"></div><li> <i class="fa fa-check-square-o fa-fw"></i> <a class="off" href="https://github.com/gavinha/TitanCNA/commit/e46683c75d28961668d10882294ea0ab53975685"> 26 Jun 2018 - <span class="text-gray">fixed bug in getSubcloneProfiles when sample has no CNA or too noisy #22</span> </a></li><div class="smallspacer"></div><li> <i class="fa fa-check-square-o fa-fw"></i> <a class="off" href="https://github.com/gavinha/TitanCNA/commit/d2687ea377ba2c59005b66fca5fc4fd6a9dec4c6"> 24 May 2018 - <span class="text-gray">fixed bug in selectSolution.R [ci skip]</span> </a></li><div class="smallspacer"></div><li> <i class="fa fa-check-square-o fa-fw"></i> <a class="off" href="https://github.com/gavinha/TitanCNA/commit/1b0d954293435c0fdd51fce769eba01b3164928c"> 11 Apr 2018 - <span class="text-gray">version bump v1.17.2</span> </a></li><div class="smallspacer"></div><li> <i class="fa fa-check-square-o fa-fw"></i> <a class="off" href="https://github.com/gavinha/TitanCNA/commit/7fc4fcc3192aa975a480ed360db0db36d46a6ae6"> 11 Apr 2018 - <span class="text-gray">include function getHaplotypesFromVCF [ci skip]</span> </a></li><div class="smallspacer"></div><li> <i class="fa fa-check-square-o fa-fw"></i> <a class="off" href="https://github.com/gavinha/TitanCNA/commit/635e89acd2fa1b75adca3c7438aa71c8ef0e8586"> 05 Apr 2018 - <span class="text-gray">remove redundant ichorCNA options [ci skip]</span> </a></li></ul></div><div class="spacer"></div><div class="smallhead"> Pages</div><div class="pad-left smallnote"><ul class="list-unstyled"><div class="smallspacer"></div><li> <a class="off" href="/projects/TitanCNA/scripts/R_scripts/"> <i class="fa fa-folder-open-o"></i> / scripts / R_scripts / </a></li><div class="smallspacer"></div><li> <a class="off" href="/projects/TitanCNA/scripts/snakemake/"> <i class="fa fa-folder-open-o"></i> / scripts / snakemake / </a></li><div class="smallspacer"></div><li> <a class="off" href="/projects/TitanCNA/scripts/TenX_scripts/"> <i class="fa fa-folder-open-o"></i> / scripts / TenX_scripts / </a></li><div class="smallspacer"></div><li> <a class="off" href="/projects/TitanCNA/scripts/"> <i class="fa fa-folder-open-o"></i> / scripts / </a></li><div class="smallspacer"></div><li> <a class="off" href="/projects/TitanCNA/"> <i class="fa fa-folder-open-o"></i> / </a></li></ul></div><div class="spacer"></div></div><div class="col-md-8"><div class="post"><p><h2 id="x-genomics-whole-genome-sequencing-analysis">2. 10X Genomics Whole Genome Sequencing Analysis</h2><p>R script (<code class="highlighter-rouge">titanCNA_v1.15.0_TenX.R</code>) for running TitanCNA analysis on sequencing data with haplotype phasing such as 10X Genomics whole genome data</p><p><strong>Input files</strong></p><ol><li>GC-corrected, normalized read coverage using the HMMcopy suite<br/> a. Get molecule coverage by counting unique barcodes in non-overlapping bins. To do this, you will need the following installed:<ul><li><a href="https://github.com/walaj/bxtools#tile">bxtools</a></li><li>samtools To extract the molecule coverage at 10kb bins across chromosome 1 for both tumour and normal samples:<br/> <code class="highlighter-rouge"> samtools view -h -F 0x4 -q 60 tumour.bam 1 | bxtools tile - -b chr1.10kb.bed &gt; tumour_bxTile/chr1.10kb.bxTile.bed </code> The chromosome bed files for 10kb bins is provided in <code class="highlighter-rouge">TenX_scripts/data/10kb.bed.tar.gz</code>. Just <code class="highlighter-rouge">tar xvzf 10kb.bed.tar.gz</code> to extract the files.</li></ul><p>b. Normalize the molecule coverage<br/> This analysis uses ichorCNA to correct molecular coverage by GC and mappability biases.<br/> See https://github.com/broadinstitute/ichorCNA/tree/master/scripts</p><p>``` # from the command line</p><blockquote><p>Rscript TenX_scripts/getMoleculeCoverage.R –help Usage: TenX_scripts/getMoleculeCoverage.R [options]</p></blockquote><p>Options: -t TUMORBXDIR, –tumorBXDir=TUMORBXDIR Path to directory containing tumor bed files for each chromosome containing BX tags.</p><div class="highlighter-rouge"><pre class="highlight"><code>     -n NORMALBXDIR, --normalBXDir=NORMALBXDIR
             Path to directory containing normal bed files for each chromosome containing BX tags.

     --minReadsPerBX=MINREADSPERBX
             Minimum number of reads per barcode.

     --gcWig=GCWIG
             Path to GC-content WIG file; Required

     --libdir=LIBDIR
             Script library path to include new or modified source R files (optional). Default: [NULL]

     (... plus other options)
</code></pre></div><p>``` Here is an example</p><p><code class="highlighter-rouge"> Rscript getMoleculeCoverage.R --id test \ --tumorBXDir tumour_bxTile/ --normalBXDir normal_bxTile/ --minReadsPerBX 2 \ --chrs "c(1:22, \"X\")" --outDir ../ \ --centromere TenX_scripts/data/GRCh37.p13_centromere_UCSC-gapTable.txt \ --gcWig ichorCNA/inst/extdata/gc_hg38_1000kb.wig \ --libdir TitanCNA/R/ </code></p></li><li><p>Tumour sample allelic read counts at heterozygous SNPs (identifed from the matched normal sample).<br/> a. Identify the phased heterozygous SNPs (excluding indels) from the LongRanger 2.1 analysis on the matched normal sample.<br/> Use this R script from the command line to process the phased variant file (<code class="highlighter-rouge">*phased_variants.vcf.gz</code>) and extracts heterozygous sites after filtering. The output VCF file suffix <code class="highlighter-rouge">*_phasedHets.vcf</code></p><p>``` # from the command line</p><blockquote><p>Rscript TenX_scripts/getPhasedHETSitesFromLLRVCF.R –help Usage: Rscript getPhasedHETSitesFromLLRVCF.R [options]</p></blockquote><p>Options: -i INVCF, –inVCF=INVCF LongRanger 2.1 phased variant result VCF file; typically has filename suffix “*phased_variants.vcf.gz”. [Required]</p><div class="highlighter-rouge"><pre class="highlight"><code>     -q MINQUALITY, --minQuality=MINQUALITY
             Heterozygous variants with QUAL greater than or equal to this value are considered. [Default: 100]

     -d MINDEPTH, --minDepth=MINDEPTH
             Heterozygous variants with read depth greater than or equal to this value are considered. [Default: 10]

     -v MINVAF, --minVAF=MINVAF
             Heterozygous variants with variant allele fraction or reference allele fraction greater than this value are considered. [Default: 0.25]

     -o OUTVCF, --outVCF=OUTVCF
             Output VCF file with suffix "_phasedHets.vcf" [Required]

     -h, --help
             Show this help message and exit  ```    b. Extract allelic read counts from the tumour sample      Using this python script, extract the allele read counts at the heterozygous SNP sites identified in the previous step.  This script is meant to be run per chromosome so that users can parallelize this step.  Users will need to combine the chromosome files into a single tab-delimited file for input into the R script.
</code></pre></div><p>``` # from the command line # run per chromosome # The <code class="highlighter-rouge">pysam</code> library will need to be installed in python.</p><p>usage: python TenX_scripts/getTumourAlleleCountsAtHETSites.py $chr $phasedHetsVCF $bam $refFasta $baseQuality $mapQuality $vcfQuality &gt; $tumCountsFile</p><p># $chr - Chromosome to analyze (e.g. 1) # $phasedHetsVCF - Output VCF file from getPhasedHETSitesFromLLRVCF.R # $bam - BAM file path of the tumour sample # $refFasta - File path to the reference FASTA file # $baseQuality - Minimum base quality to include in counts # $mapuality - Minimum mapping quality of reads to include in counts # $vcfQuality - Exclude HET sites with lower QUAL than this value # $tumCountsFile - Output tab-delimited file ```</p></li></ol><p><strong>Running the R script</strong></p><ol><li>Look at the usage of the R script</li></ol><p>``` # from the command line</p><blockquote><p>Rscript titanCNA_v1.15.0_TenX.R –help Usage: Rscript titanCNA_v1.15.0_TenX.R [options]</p></blockquote><p>Options: –id=ID Sample ID</p><div class="highlighter-rouge"><pre class="highlight"><code>    --hetFile=HETFILE
            File containing allelic read counts at HET sites. (Required)

    (... other arguments similar to titanCNA.R)

    --alleleModel=ALLELEMODEL
            Emission density to use for allelic input data (binomial or Gaussian). [Default: binomial]

    --alphaR=ALPHAR
            Hyperparaemter on the Gaussian variance for allelic fraction data; used if --alleleModel="Gaussian". [Defaule: 5000]

    --haplotypeBinSize=HAPLOTYPEBINSIZE
            Bin size for summarizing phased haplotypes. [Default: 100000]

    --phaseSummarizeFun=PHASESUMMARIZEFUN
            Strategy to summarize specified haplotype bins: mean, sum, SNP. [Default: sum]   ```   The specific arguments that are different compared to `titanCNA.R` are listed above.  `HETFILE` should be the output of `getTumourAlleleCountsAtHETSites.py`.
</code></pre></div><ol><li>Example usage of R script</li></ol><p><code class="highlighter-rouge"> # normalized coverage file: test.cn.txt # allelic read count file from getTumourAlleleCountsAtHETSites.py: test.het.txt Rscript titanCNA.R --id test --hetFile test.het.txt --cnFile test.cn.txt \ --numClusters 1 --numCores 1 --normal_0 0.5 --ploidy_0 2 \ --chrs "c(1:22, \"X\")" --estimatePloidy TRUE --outDir ./ \ --haplotypeBinSize 1e5 --phaseSummarizeFun sum --alleleModel Gaussian --alphaR 5000 </code></p><ol><li>Running TitanCNA for multiple restarts and model selection<br/> Use the same approach as for Step 3 of <a href="#wgs">Standard Whole Genome/Exome Sequencing Analysis</a>.</li></ol></div></div><div class="col-md-1"></div></div><div class="bigspacer"></div><div class="hugespacer"></div></div><div id="footer"> <nav class="navbar navbar-default navbar-fixed-bottom" role="navigation"><div class="container"><div class="navbar-footer"> <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar-collapse-element" aria-expanded="false"> <span class="sr-only">Toggle navigation</span> <span class="icon-bar"></span> <span class="icon-bar"></span> <span class="icon-bar"></span> </button></div> <!object class="logofooter navbar-brand" data="/images/fhcrc-logo.png" type="image/png"><!/object><div class="collapse navbar-collapse" id="navbar-collapse-element"><ul class="nav navbar-nav navbar-right navbar-tiny"><li><a class="off" href="/misc/about/">About</a></li><li><a class="off" href="/misc/contact/">Contact</a></li></ul><ul class="nav navbar-text"></ul></div></div> </nav></div></body></html>