<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"/><title>TitanCNA/scripts/snakemake/</title><meta name="viewport" content="width=device-width, initial-scale=1.0"><link rel="alternate" type="application/rss+xml" title="Ha Lab" href="http://gavinha.github.io/feed.xml"><link href="/css/style.css" rel="stylesheet" media="screen"> <script>!function(e){var t,a={kitId:"xeu8jut",scriptTimeout:3e3},c=e.documentElement,i=setTimeout(function(){c.className=c.className.replace(/\bwf-loading\b/g,"")+" wf-inactive"},a.scriptTimeout),n=e.createElement("script"),o=!1,s=e.getElementsByTagName("script")[0];c.className+=" wf-loading",n.src="//use.typekit.net/"+a.kitId+".js",n.async=!0,n.onload=n.onreadystatechange=function(){if(t=this.readyState,!(o||t&&"complete"!=t&&"loaded"!=t)){o=!0,clearTimeout(i);try{Typekit.load(a)}catch(e){}}},s.parentNode.insertBefore(n,s)}(document)</script> <script>!function(e,a,n,t,c,o,s){e.GoogleAnalyticsObject=c,e[c]=e[c]||function(){(e[c].q=e[c].q||[]).push(arguments)},e[c].l=1*new Date,o=a.createElement(n),s=a.getElementsByTagName(n)[0],o.async=1,o.src=t,s.parentNode.insertBefore(o,s)}(window,document,"script","//www.google-analytics.com/analytics.js","ga"),ga("create","UA-26244371-2","bedford.io"),ga("send","pageview")</script> <script src="/js/jquery.min.js"></script> <!link href="css/bootstrap.min.css" rel="stylesheet"> <script src="/js/transition.js"></script><script src="/js/collapse.js"></script><script src="/js/less.min.js"></script></head><body><div id="header"> <nav class="navbar navbar-default navbar-fixed-top" role="navigation"><div class="container"><div class="navbar-header"> <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar-collapse-element" aria-expanded="false"> <span class="sr-only">Toggle navigation</span> <span class="icon-bar"></span> <span class="icon-bar"></span> <span class="icon-bar"></span> </button> <!object class="logo navbar-brand" data="/images/logo.svg" type="image/svg+xml"><!/object> <a class="navbar-brand" href="/">Ha Lab</a></div><div class="collapse navbar-collapse" id="navbar-collapse-element"><ul class="nav navbar-nav navbar-right"><li> <a href="/news/">News</a></li><li> <a href="/research/">Research</a></li><li> <a href="/people/">People</a></li><li> <a href="/papers/">Papers</a></li><li> <a href="/projects/">Software</a></li><li><li><a class="off" href="/join/join/">Join</a></li></li></ul></div></div> </nav></div><div class="container"><div class="hugespacer"></div><div class="bigspacer"></div><div class="row"><div class="col-md-12"><div class="title titlebox"><ol class="breadcrumb"><li><a class="off" href="/projects/TitanCNA/">TitanCNA</a></li><li><a class="off" href="/projects/TitanCNA/scripts/">scripts</a></li><li>snakemake</li></ol></div><p><div class="head"> Analysis of subclonal copy number alterations (CNA) and loss of heterozygosity (LOH) in cancer</div></div></div><div class="bigspacer"></div><div class="row"><div class="col-md-3"><div class="bigspacer"></div><div class="smallhead"> Source code</div><div class="pad-left note"><div class="smallspacer"></div> <i class="fa fa-cog fa-fw"></i> <a class="off" href="https://github.com/gavinha/TitanCNA">github.com/gavinha/TitanCNA</a></div><div class="bigspacer"></div><div class="smallhead"> Contributors</div><div class="pad-left note"><div class="smallspacer"></div><div> <a class="off" href="https://github.com/gavinha"> <img class="pull-left avatar" src="https://avatars1.githubusercontent.com/u/9221688?v=4&s=50"><div class="handlebox" style="padding-left:5px""> gavinha</div> </a></div><div class="smallspacer"></div><div> <a class="off" href="https://github.com/lbeltrame"> <img class="pull-left avatar" src="https://avatars0.githubusercontent.com/u/25192?v=4&s=50"><div class="handlebox" style="padding-left:5px""> lbeltrame</div> </a></div><div class="smallspacer"></div><div> <a class="off" href="https://github.com/dtenenba"> <img class="pull-left avatar" src="https://avatars0.githubusercontent.com/u/2286826?v=4&s=50"><div class="handlebox" style="padding-left:5px""> dtenenba</div> </a></div><div class="smallspacer"></div><div> <a class="off" href="https://github.com/hpages"> <img class="pull-left avatar" src="https://avatars1.githubusercontent.com/u/8810451?v=4&s=50"><div class="handlebox" style="padding-left:5px""> hpages</div> </a></div><div class="smallspacer"></div><div> <a class="off" href="https://github.com/mrjc42"> <img class="pull-left avatar" src="https://avatars3.githubusercontent.com/u/2029097?v=4&s=50"><div class="handlebox" style="padding-left:5px""> mrjc42</div> </a></div></div><div class="bigspacer"></div><div class="smallhead"> Latest commits</div><div class="pad-left smallnote"><ul class="list-unstyled"><div class="smallspacer"></div><li> <i class="fa fa-check-square-o fa-fw"></i> <a class="off" href="https://github.com/gavinha/TitanCNA/commit/e46683c75d28961668d10882294ea0ab53975685"> 26 Jun 2018 - <span class="text-gray">fixed bug in getSubcloneProfiles when sample has no CNA or too noisy #22</span> </a></li><div class="smallspacer"></div><li> <i class="fa fa-check-square-o fa-fw"></i> <a class="off" href="https://github.com/gavinha/TitanCNA/commit/d2687ea377ba2c59005b66fca5fc4fd6a9dec4c6"> 24 May 2018 - <span class="text-gray">fixed bug in selectSolution.R [ci skip]</span> </a></li><div class="smallspacer"></div><li> <i class="fa fa-check-square-o fa-fw"></i> <a class="off" href="https://github.com/gavinha/TitanCNA/commit/1b0d954293435c0fdd51fce769eba01b3164928c"> 11 Apr 2018 - <span class="text-gray">version bump v1.17.2</span> </a></li><div class="smallspacer"></div><li> <i class="fa fa-check-square-o fa-fw"></i> <a class="off" href="https://github.com/gavinha/TitanCNA/commit/7fc4fcc3192aa975a480ed360db0db36d46a6ae6"> 11 Apr 2018 - <span class="text-gray">include function getHaplotypesFromVCF [ci skip]</span> </a></li><div class="smallspacer"></div><li> <i class="fa fa-check-square-o fa-fw"></i> <a class="off" href="https://github.com/gavinha/TitanCNA/commit/635e89acd2fa1b75adca3c7438aa71c8ef0e8586"> 05 Apr 2018 - <span class="text-gray">remove redundant ichorCNA options [ci skip]</span> </a></li></ul></div><div class="spacer"></div><div class="smallhead"> Pages</div><div class="pad-left smallnote"><ul class="list-unstyled"><div class="smallspacer"></div><li> <a class="off" href="/projects/TitanCNA/scripts/R_scripts/"> <i class="fa fa-folder-open-o"></i> / scripts / R_scripts / </a></li><div class="smallspacer"></div><li> <a class="off" href="/projects/TitanCNA/scripts/snakemake/"> <i class="fa fa-folder-open-o"></i> / scripts / snakemake / </a></li><div class="smallspacer"></div><li> <a class="off" href="/projects/TitanCNA/scripts/TenX_scripts/"> <i class="fa fa-folder-open-o"></i> / scripts / TenX_scripts / </a></li><div class="smallspacer"></div><li> <a class="off" href="/projects/TitanCNA/scripts/"> <i class="fa fa-folder-open-o"></i> / scripts / </a></li><div class="smallspacer"></div><li> <a class="off" href="/projects/TitanCNA/"> <i class="fa fa-folder-open-o"></i> / </a></li></ul></div><div class="spacer"></div></div><div class="col-md-8"><div class="post"><p><h1 id="snakemake-workflow-for-titan"><em>Snakemake workflow for TITAN</em></h1><h2 id="description">Description</h2><p>This workflow will run the TITAN a set of tumour-normal pairs, starting from the BAM files and generating TitanCNA outputs. It will also perform model selection at the end of the workflow to choose the optimal ploidy and clonal cluster solutions.</p><h2 id="requirements">Requirements</h2><h3 id="software-packages-or-libraries">Software packages or libraries</h3><ul><li>R-3.3<ul><li>TitanCNA (v1.15.0) - TitanCNA imports: GenomicRanges, dplyr, data.table, doMC</li><li><a href="https://github.com/broadinstitute/ichorCNA">ichorCNA</a> (v0.1.0)</li><li>HMMcopy</li><li>optparse</li><li>stringr</li><li>SNPchip</li></ul></li><li>Python 3.4<ul><li>snakemake-3.12.0</li><li>PySAM-0.11.2.1</li><li>PyYAML-3.12</li></ul></li><li>samtools v1.3.1</li><li>bcftools v1.1</li><li><a href="http://shahlab.ca/projects/hmmcopy_utils/">HMMcopy Suite</a>.<br/> -In particular, <code class="highlighter-rouge">readCounter</code> is used.</li></ul><h3 id="scriptsexecutables">Scripts/executables</h3><ul><li>readCounter (C++ executable; HMMcopy Suite)</li><li>ichorCNA.R (ichorCNA tool for normalizing and correcting read coverage)</li><li><a href="https://github.com/gavinha/TitanCNA/tree/master/scripts/snakemake/code/countPysam.py">countPysam.py</a> (generates input allele counts)</li><li><a href="https://github.com/gavinha/TitanCNA/tree/master/scripts/snakemake/../R_scripts/titanCNA.R">titanCNA.R</a> (main R script to run TitanCNA)</li><li><a href="https://github.com/gavinha/TitanCNA/tree/master/scripts/snakemake/../R_scripts/selectSolution.R">selectSolution.R</a> (R script to select optimal solution for each sample)</li></ul><h2 id="tumour-normal-sample-list">Tumour-Normal sample list</h2><p>The list of tumour-normal paired samples should be defined in a YAML file. See <code class="highlighter-rouge">config/samples.yaml</code> for an example. Both fields <code class="highlighter-rouge">samples</code> and <code class="highlighter-rouge">pairings</code> must to be provided. <code class="highlighter-rouge">pairings</code> key must match the tumour sample while the value must match the normal sample.</p><div class="highlighter-rouge"><pre class="highlight"><code>samples:
  tumor_sample_1:  /path/to/bam/tumor.bam
  normal_sample_1:  /path/to/bam/normal.bam


pairings:
  tumor_sample_1:  normal_sample_1
</code></pre></div><h2 id="snakefiles">snakefiles</h2><ol><li><code class="highlighter-rouge">ichorCNA.snakefile</code></li><li><code class="highlighter-rouge">getAlleleCounts.snakefile</code></li><li><code class="highlighter-rouge">TitanCNA.snakefile</code></li></ol><p>Invoking the full snakemake workflow for TITAN</p><div class="highlighter-rouge"><pre class="highlight"><code># show commands and workflow
snakemake -s TitanCNA.snakefile -np
# run the workflow locally using 5 cores
snakemake -s TitanCNA.snakefile --cores 5
# run the workflow on qsub using a maximum of 50 jobs. Broad UGER cluster parameters can be set directly in config/cluster.sh.
snakemake -s TitanCNA.snakefile --cluster-sync "qsub" -j 50 --jobscript config/cluster.sh
</code></pre></div><p>This will also run both <code class="highlighter-rouge">ichorCNA.snakefile</code> and <code class="highlighter-rouge">getAlleleCounts.snakefile</code> which generate the necessary inputs for <code class="highlighter-rouge">TitanCNA.snakfile</code>.</p><p><code class="highlighter-rouge">ichorCNA.snakefile</code> and <code class="highlighter-rouge">getAlleleCounts.snakefile</code> can also be invoked separately. If only one but not both results are needed, then you can invoke the snakefiles independently.</p><div class="highlighter-rouge"><pre class="highlight"><code>snakemake -s ichorCNA.snakefile --cores 5
# OR
snakemake -s getAlleleCounts.snakefile --cores 5
</code></pre></div></div></div><div class="col-md-1"></div></div><div class="bigspacer"></div><div class="hugespacer"></div></div><div id="footer"> <nav class="navbar navbar-default navbar-fixed-bottom" role="navigation"><div class="container"><div class="navbar-footer"> <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar-collapse-element" aria-expanded="false"> <span class="sr-only">Toggle navigation</span> <span class="icon-bar"></span> <span class="icon-bar"></span> <span class="icon-bar"></span> </button></div> <!object class="logofooter navbar-brand" data="/images/fhcrc-logo.png" type="image/png"><!/object><div class="collapse navbar-collapse" id="navbar-collapse-element"><ul class="nav navbar-nav navbar-right navbar-tiny"><li><a class="off" href="/misc/about/">About</a></li><li><a class="off" href="/misc/contact/">Contact</a></li></ul><ul class="nav navbar-text"></ul></div></div> </nav></div></body></html>